fieldtracing/fieldtracing.cpp

```diff
@@ -0,0 +1,1701 @@
+      std::vector<uint64_t> maxCellFWStepCounter(globalDccrgSize, 0);
+      std::vector<uint64_t> maxCellBWStepCounter(globalDccrgSize, 0);
+      std::vector<Real> reducedCellFWTracingStepSize(globalDccrgSize);
+      std::vector<Real> reducedCellBWTracingStepSize(globalDccrgSize);
+      
+      phiprof::start("first-loop");
+      for(int n=0; n<globalDccrgSize; n++) {
+         const CellID id = allDccrgCells[n];
+         cellFWTracingCoordinates.at(n) = mpiGrid.get_center(id);
+         cellBWTracingCoordinates.at(n) = cellFWTracingCoordinates.at(n);
+         
+         if(mpiGrid.is_local(id)) {
+            if((mpiGrid[id]->sysBoundaryFlag != sysboundarytype::NOT_SYSBOUNDARY)
+               || cellFWTracingCoordinates[n][0] > P::xmax - 4*P::dx_ini
+               || cellFWTracingCoordinates[n][0] < P::xmin + 4*P::dx_ini
+               || cellFWTracingCoordinates[n][1] > P::ymax - 4*P::dy_ini
+               || cellFWTracingCoordinates[n][1] < P::ymin + 4*P::dy_ini
+               || cellFWTracingCoordinates[n][2] > P::zmax - 4*P::dz_ini
+               || cellFWTracingCoordinates[n][2] < P::zmin + 4*P::dz_ini
+            ) {
+               cellNeedsContinuedFWTracing[n] = 0;
+               cellNeedsContinuedBWTracing[n] = 0;
+               cellFWTracingCoordinates[n] = {0,0,0};
+               cellBWTracingCoordinates[n] = {0,0,0};
+               cellFWTracingStepSize[n] = 0;
+               cellBWTracingStepSize[n] = 0;
+            }
+         }
+      }
+      phiprof::stop("first-loop");
+      MPI_Allreduce(cellNeedsContinuedFWTracing.data(), reducedCellNeedsContinuedFWTracing.data(), globalDccrgSize, MPI_INT, MPI_MIN, MPI_COMM_WORLD);
+      MPI_Allreduce(cellNeedsContinuedBWTracing.data(), reducedCellNeedsContinuedBWTracing.data(), globalDccrgSize, MPI_INT, MPI_MIN, MPI_COMM_WORLD);
+      if(sizeof(Real) == sizeof(double)) {
+         MPI_Allreduce(cellFWTracingStepSize.data(), reducedCellFWTracingStepSize.data(), globalDccrgSize, MPI_DOUBLE, MPI_MIN, MPI_COMM_WORLD);
+         MPI_Allreduce(cellBWTracingStepSize.data(), reducedCellBWTracingStepSize.data(), globalDccrgSize, MPI_DOUBLE, MPI_MIN, MPI_COMM_WORLD);
+      } else {
+         MPI_Allreduce(cellFWTracingStepSize.data(), reducedCellFWTracingStepSize.data(), globalDccrgSize, MPI_FLOAT, MPI_MIN, MPI_COMM_WORLD);
+         MPI_Allreduce(cellBWTracingStepSize.data(), reducedCellBWTracingStepSize.data(), globalDccrgSize, MPI_FLOAT, MPI_MIN, MPI_COMM_WORLD);
+      }
+      cellNeedsContinuedFWTracing = reducedCellNeedsContinuedFWTracing;
+      cellNeedsContinuedBWTracing = reducedCellNeedsContinuedBWTracing;
+      cellFWTracingStepSize = reducedCellFWTracingStepSize;
+      cellBWTracingStepSize = reducedCellBWTracingStepSize;
+      bool anyCellNeedsTracing;
+      
+      // Fieldline tracing function
+      TracingFieldFunction tracingFullField = [&perBGrid, &dPerBGrid, &technicalGrid](std::array<Real,3>& r, const bool alongB, std::array<Real,3>& b)->bool {
```

Comment:
This is, again, basically the same tracing function. As far as I can tell, it only differs in its behaviour when tracing outside the domain or inside the ionosphere.